File: http/PHP/lib/httpclient.php
<?php
declare(strict_types=1);
/*
 * Copyright Core DF — Apache License 2.0
 * Generic HTTP client helpers for Core Auto step scripts.
 */
require_once __DIR__ . '/result.php';

final class Httpclient
{
    private static function parseBody(string $raw): mixed
    {
        if ($raw === '') {
            return null;
        }
        $parsed = json_decode($raw, true);
        return json_last_error() === JSON_ERROR_NONE ? $parsed : $raw;
    }

    /** @param array<string,string>|null $headers */
    private static function request(string $method, string $url, ?array $headers = null, ?string $body = null, ?array $params = null): array
    {
        if ($params) {
            $sep = str_contains($url, '?') ? '&' : '?';
            $url .= $sep . http_build_query($params);
        }
        $ch = curl_init($url);
        if ($ch === false) {
            return CoreautoResult::transportError();
        }
        $hdrs = [];
        foreach ($headers ?? [] as $k => $v) {
            $hdrs[] = "{$k}: {$v}";
        }
        curl_setopt_array($ch, [
            CURLOPT_CUSTOMREQUEST => $method,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_HTTPHEADER => $hdrs,
            CURLOPT_TIMEOUT => 60,
        ]);
        if ($body !== null) {
            curl_setopt($ch, CURLOPT_POSTFIELDS, $body);
        }
        $raw = curl_exec($ch);
        $code = (int) curl_getinfo($ch, CURLINFO_RESPONSE_CODE);
        $err = curl_error($ch);
        curl_close($ch);
        if ($raw === false) {
            return CoreautoResult::transportError($err ?: 'inaccessible');
        }
        $parsed = self::parseBody($raw);
        if ($code >= 400) {
            return ['status_code' => $code, 'error' => $parsed ?? 'inaccessible'];
        }
        return ['status_code' => $code, 'body' => $parsed];
    }

    public static function Get(string $url, ?array $headers = null, ?array $params = null): array
    {
        return self::request('GET', $url, $headers, null, $params);
    }

    public static function Post(string $url, mixed $json_body = null, ?string $data = null, ?array $headers = null): array
    {
        $hdrs = $headers ?? [];
        $body = null;
        if ($json_body !== null) {
            $hdrs['Content-Type'] ??= 'application/json';
            $body = json_encode($json_body, JSON_THROW_ON_ERROR);
        } elseif ($data !== null) {
            $body = $data;
        }
        return self::request('POST', $url, $hdrs, $body);
    }

    public static function Put(string $url, mixed $json_body = null, ?array $headers = null): array
    {
        $hdrs = $headers ?? [];
        $body = null;
        if ($json_body !== null) {
            $hdrs['Content-Type'] ??= 'application/json';
            $body = json_encode($json_body, JSON_THROW_ON_ERROR);
        }
        return self::request('PUT', $url, $hdrs, $body);
    }

    public static function Delete(string $url, ?array $headers = null): array
    {
        return self::request('DELETE', $url, $headers);
    }
}

File: http/Dart/lib/httpclient.dart
// Copyright Core DF — Apache License 2.0
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'result.dart';

class Httpclient {
  static dynamic _parseBody(String raw) {
    if (raw.isEmpty) return null;
    try { return jsonDecode(raw); } catch (_) { return raw; }
  }

  static Future<Map<String, dynamic>> _request(String method, String url,
      {Map<String, String>? headers, String? body, Map<String, String>? params}) async {
    var u = Uri.parse(url);
    if (params != null && params.isNotEmpty) {
      u = u.replace(queryParameters: {...u.queryParameters, ...params});
    }
    final hdrs = Map<String, String>.from(headers ?? {});
    try {
      late http.Response resp;
      if (method == 'GET') resp = await http.get(u, headers: hdrs);
      else if (method == 'DELETE') resp = await http.delete(u, headers: hdrs);
      else if (method == 'PUT') resp = await http.put(u, headers: hdrs, body: body);
      else resp = await http.post(u, headers: hdrs, body: body);
      final parsed = _parseBody(resp.body);
      if (resp.statusCode >= 400) {
        return {'status_code': resp.statusCode, 'error': parsed ?? 'inaccessible'};
      }
      return {'status_code': resp.statusCode, 'body': parsed};
    } catch (e) {
      return CoreautoResult.transportError(e.toString());
    }
  }

  static Future<Map<String, dynamic>> Get(String url,
          {Map<String, String>? headers, Map<String, String>? params}) =>
      _request('GET', url, headers: headers, params: params);

  static Future<Map<String, dynamic>> Post(String url,
          {Object? jsonBody, String? data, Map<String, String>? headers}) async {
    final hdrs = Map<String, String>.from(headers ?? {});
    String? body;
    if (jsonBody != null) {
      hdrs.putIfAbsent('Content-Type', () => 'application/json');
      body = jsonEncode(jsonBody);
    } else if (data != null) body = data;
    return _request('POST', url, headers: hdrs, body: body);
  }

  static Future<Map<String, dynamic>> Put(String url,
          {Object? jsonBody, Map<String, String>? headers}) async {
    final hdrs = Map<String, String>.from(headers ?? {});
    String? body;
    if (jsonBody != null) {
      hdrs.putIfAbsent('Content-Type', () => 'application/json');
      body = jsonEncode(jsonBody);
    }
    return _request('PUT', url, headers: hdrs, body: body);
  }

  static Future<Map<String, dynamic>> Delete(String url,
          {Map<String, String>? headers}) =>
      _request('DELETE', url, headers: headers);
}
